### ISSUE TITLE
Add FocusNode to Textfield in buildEditWidget

### ISSUE THREAD

**Issue by abdulrehmananwar** [syncfusion/flutter-widgets#694]
i want to add FocusNode to TextField in buildEditWidget butt when i add its not working properly
    _focusNode.addListener(() async{
    if (!_focusNode.hasFocus) {print('done');}});

**Comment by Yuvaraj-Gajaraj** [comment]
Hi @abdulrehmananwar,

Greetings from Syncfusion. We suspect that you are facing issues in Flutter native widget (TextField) and not using this in our Flutter widget. In case you are using TextField in any of our widgets, kindly get back to us in which widget you are using and share some code and provide scenarios for your issue. In case your scenario is not with our widgets, then you can post your query in StackOverflow or log an issue report to the Flutter team using the link below.

https://github.com/flutter/flutter

Regards,
Yuvaraj.

**Comment by abdulrehmananwar** [comment]
this issue is in syncfusion_flutter_datagrid gridcell

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @abdulrehmananwar,

As we said in the previous update https://github.com/syncfusion/flutter-widgets/issues/672#issuecomment-1103830255 , you won't be able to process based on the TextField focus node. Because we get the buildEditWidget as a widget and the datagrid does not process the focus based on that. The DataGrid progressed; the focus was based on the cell focus. If you tap any other cell, the focus immediately goes to the DataGrid.

Regards,
Tamilarasan

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @abdulrehmananwar,

As we said in this update https://github.com/syncfusion/flutter-widgets/issues/672#issuecomment-1103830255, we have checked all the possible ways to achieve your requirement with the workaround. We couldn’t achieve it at the sample level itself, and there is no alternative solution for this apart from that feature, i.e., making the onSubmit method async.

Regards,
Tamilarasan

**Comment by abdulrehmananwar** [comment]
on Submit  event focus goes to the next field. when i call async function. how to hold focus on that cell uptill finish function. if function return false value then hold focus same sum.

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @abdulrehmananwar,

We have planned to implement the feature to call future functions while editing the cell. If the value returns false, the cell focus doesn't move to another cell. It remains in the same cell until that point in time that we mentioned. Please check this code snippet. Can you please confirm whether this meets your requirements?

```dart
  @override
  Future<bool> canSubmitCell(DataGridRow dataGridRow,
      RowColumnIndex rowColumnIndex, GridColumn column) async {
    generateProductList();
    await Future.delayed(const Duration(seconds: 3));

    if (!productlist.any((element) => element.customerID == newCellValue)) {
      return false;
    }

    return true;
  }

Future generateProductList() async {
  var response = await http.get(Uri.parse(
      'https://ej2services.syncfusion.com/production/web-services/api/Orders'));
  var list = json.decode(response.body).cast<Map<String, dynamic>>();
  productlist =
      await list.map<_Product>((json) => _Product.fromJson(json)).toList();
  return productlist;
}


```

Regards,
Tamilarasan

**Comment by abdulrehmananwar** [comment]
this is perfect i already do the same few weeks ago. butt if any widget outside the datagrid like a save button the that button is pressable. i want no other widget should work outside untill value is exist. if we can implement focus node to textfield the we can able to resolve this issue.

**Comment by ashok-kuvaraja** [comment]
Hi @abdulrehmananwar,

You need to disable and enable the save button based on the CanSubmitCell return value. You can easily achieve your requirement in this implementation. You have to listen to the new value inside the CanSubmitCell and process the button to enable and disable operation by calling the state's setState method. Please find the code snippet below,

#### DataGrid Sample
```dart
  @override
  void initState() {
    super.initState();
    dataGridSource = DealerDataGridSource(setState);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Expanded(
              child: SfDataGrid(
            source: dataGridSource,
            allowEditing: true,
            navigationMode: GridNavigationMode.cell,
            selectionMode: SelectionMode.single,
            columns: <GridColumn>[
              GridColumn(
                  columnName: 'Product No',
                  width: 110,
                  label: Container(
                    padding: const EdgeInsets.all(8.0),
                    alignment: Alignment.centerRight,
                    child: const Text(
                      'Product No',
                      overflow: TextOverflow.ellipsis,
                    ),
                  )),
              GridColumn(
                  columnName: 'Dealer Name',
                  width: 110,
                  label: Container(
                    padding: const EdgeInsets.all(8.0),
                    alignment: Alignment.centerLeft,
                    child: const Text(
                      'Dealer Name',
                      overflow: TextOverflow.ellipsis,
                    ),
                  )),
              GridColumn(
                  columnName: 'Dealer ID',
                  width: 110,
                  label: Container(
                    padding: const EdgeInsets.all(8.0),
                    alignment: Alignment.centerRight,
                    child: const Text(
                      'Dealer ID',
                      overflow: TextOverflow.ellipsis,
                    ),
                  )),
              GridColumn(
                  columnName: 'Price',
                  label: Container(
                    padding: const EdgeInsets.all(8.0),
                    alignment: Alignment.centerRight,
                    child: const Text(
                      'Price',
                      overflow: TextOverflow.ellipsis,
                    ),
                  )),
            ],
          )),
          Padding(
            padding: EdgeInsets.all(20.0),
            child: MaterialButton(
                color: Colors.blue,
                child: Text('Save'),
                onPressed: dataGridSource.isButtonEnabled ? () {} : null),
          )
        ],
      ),
    );
  }
```
  #### DataGridSource

```dart
class DealerDataGridSource extends DataGridSource {
  DealerDataGridSource(this.rebuild);
  
...
...
...

  // Calls the field to rebuild the DataGridSample.
  StateSetter rebuild;
  
  // Decides whether the save button should be enabled or not.
  bool isButtonEnabled = false;

  @override
  Future<bool> canSubmitCell(DataGridRow dataGridRow,
      RowColumnIndex rowColumnIndex, GridColumn column) async {
    final dynamic oldValue = dataGridRow
            .getCells()
            .firstWhereOrNull((DataGridCell dataGridCell) =>
                dataGridCell.columnName == column.columnName)
            ?.value ??
        '';

    await Future.delayed(Duration(seconds: 2));

    if (newCellValue == null || oldValue == newCellValue) {
      // To disable the save button when the old and new value is same.
      rebuild(() {
        isButtonEnabled = false;
      });
      return false;
    }

     // To enable the save button when the old and new value is not same.
    rebuild(() {
      isButtonEnabled = true;
    });
    return true;
  }
  
...
...
...

}
```

#### Demo

https://user-images.githubusercontent.com/60499432/166410171-9d8c2c25-5aa8-400a-ac64-52f6851fa7b9.mp4

If the issue is still not resolved. Please provide more details to check this issue further.

Regards,
Ashok K

**Comment by abdulrehmananwar** [comment]
is async feature on  canSubmitCell event available now?

**Comment by ashok-kuvaraja** [comment]
Hi @abdulrehmananwar,

Currently, this feature is not available in SfDataGrid. We will provide this feature and it will be rolled out in the upcoming Vol 1 SP release. We appreciate your patience until then.

Regards,
Ashok K

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @abdulrehmananwar,

Sorry for the inconvenience caused.

Due to some technical difficulties, unfortunately we couldn't provide the async support for the onCellSubmit and canSubmitCell methods in the 2022 Volume 1 SP release. We will implement this feature in our 2022 Volume 2 release, which is expected to be rolled out at the end of June 2022. We will let you know once it is published. We appreciate your patience until then.

Regards,
Tamilarasan

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @abdulrehmananwar,

Sorry for the inconvenience caused.

Due to some technical difficulties, we unfortunately couldn't provide the async support for the onCellSubmit method in the 2022 Volume 2  release. We will implement this feature in our 2022 Volume 3 release, which is expected to be rolled out by the end of September 2022. We will let you know once it is published. We appreciate your patience and understanding till then.

Regards,
Tamilarasan

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @abdulrehmananwar,

Sorry for the inconvenience caused.

Unfortunately, we couldn't provide the async support for the onCellSubmit method in the 2022 Volume 3 release. We will implement this feature in our 2022 Volume 4 release, which is expected to be rolled out by the mid of December 2022. We will let you know once it is published. We appreciate your patience and understanding till then.

Regards,
Tamilarasan

**Comment by Tamilarasan-Paranthaman** [comment]
@abdulrehmananwar,

Sorry for the inconvenience caused. Unfortunately, we couldn't provide the async support for the onCellSubmit method in the 2022 Volume 4 release. We will implement this feature in our 2023 Volume 1 release, which is expected to be rolled out by the end of March 2023. We will let you know once it is published. We appreciate your patience and understanding until then.

**Comment by Tamilarasan-Paranthaman** [comment]
@abdulrehmananwar,

We are glad to announce that our Essential Studio 2023 Volume 1 Main Release V21.1.0.35 is rolled out. The support for making the onCellSubmit method async to process the asynchronous operation on editing completion has been included in version **21.1.35**.

Pub Link: https://pub.dev/packages/syncfusion_flutter_datagrid/changelog

UG Documentation: https://help.syncfusion.com/flutter/datagrid/editing#perform-editing-asynchronously

We thank you for your support and appreciate your patience in waiting for this release. Please get in touch with us if you would require any further assistance.
